fix(obstacle_stop_planner): Add checking of point height.

lchojnack opened the pull request:
## Description

<!-- Write a brief description of this PR. -->

This PR adds checking of a point height. If a point is lower than the vehicle's height, then the point is in collision with the vehicle. This allows the EGO to drive when an obstacle is above the EGO e.g. tree branches and leaves.

Related to:
- issue https://github.com/autowarefoundation/autoware.universe/issues/2213 
- discussion https://github.com/orgs/autowarefoundation/discussions/2981

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

maxime-clem reviewed (commented):
The idea seems good, expect for the case where ego is about to drive up a slope.
I believe in that case we can encounter an obstacle that is _currently_ at a higher `z` than the vehicle height, but that will still collide with the ego vehicle if it follows its trajectory.
![image](https://user-images.githubusercontent.com/78338830/201932642-5a6c52c1-7a97-4321-9ef8-aa04057178ac.png)

What do you think ?

lchojnack commented on planning/obstacle_stop_planner/src/planner_utils.cpp:
```diff
@@ -441,8 +441,10 @@ bool withinPolygon(
       if (bg::within(point, boost_polygon)) {
-        within_points_ptr->push_back(candidate_points_ptr->at(j));
-        find_within_points = true;
+        if (candidate_points_ptr->at(j).z < vehicle_height) {
```
As I understand from the code, the function `searchPointcloudNearTrajectory()` transforms the received point cloud to the trajectory frame, and then the transformed point cloud is used to detect collision between the trajectory and a point from the point cloud (`withinPolygon()` function).

So when the point cloud and the trajectory are in the same frame, the case with a slope is taken into consideration. 
Taking your example, I draw how trajectory would look from the side view in my opinion.
![trajectory](https://user-images.githubusercontent.com/8416175/201950014-084e39c3-fac1-42ed-bb2c-2bcea0ae3f8f.png)

Is my reasoning correct or did I miss something?

badai-nguyen replied:
I also think this is good idea. To cover driving up a slope case, a dynamic z threshold might be better. How do you thing?
![image](https://user-images.githubusercontent.com/94814556/202050509-1a5d0309-62d0-46ab-8d06-24a83b1d0849.png)

maxime-clem replied:
Sorry I have not had time to check in details yet but I started testing this PR.
It is not easy to test in the simple planning simulator but I managed to create 2 situations that look wrong.

- First, obstacle points that are under the trajectory are causing the obstacle stop planner to stop.
![image](https://user-images.githubusercontent.com/78338830/202178636-add24756-5d65-4240-b903-c49237ee96a6.png)

- Second, obstacle points that are on the elevated part of the trajectory are ignored if the ego vehicle is at a similar elevation. If the ego vehicle is at a lower elevation they correctly cause a stop.
![image](https://user-images.githubusercontent.com/78338830/202180059-e390e0f0-5130-459c-a9dc-7da8cde5dfff.png)
![image](https://user-images.githubusercontent.com/78338830/202180613-ffe4b23e-4cb5-40bd-ab3b-f717ac9be5d0.png)

I think this highlight some issues with comparing the `z` values from trajectory and from the pointcloud.

lchojnack commented:
I've updated the PR. I've added a polyhedron for visualization (topic: `/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/debug/marker`) and colliding point cloud (topic: `/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/output/point_cloud/debug`). 

I've tested the PR in some scenarios, and I think it works now. Printscreens below
<img width="1920" alt="Screenshot_20221129_101853" src="https://user-images.githubusercontent.com/8416175/204495977-cedcb726-e7fd-4f84-8863-148e66ef4790.png">
<img width="1920" alt="Screenshot_20221128_153850" src="https://user-images.githubusercontent.com/8416175/204496098-4410caac-6c27-4f24-849a-0e252cf75f08.png">
<img width="1920" alt="Screenshot_20221128_153753" src="https://user-images.githubusercontent.com/8416175/204496170-168bb9d2-6d3c-43c9-86cd-e3814ec26083.png">
<img width="1920" alt="Screenshot_20221128_153701" src="https://user-images.githubusercontent.com/8416175/204496206-c54c0db5-9088-4587-a5b2-111e39c1241d.png">

maxime-clem reviewed (commented):
Thank you for the update. I tested and it is working well.

I think the main remaining issue is that the polyhedrons are not rotated to match the road inclination. Since this is probably difficult to implement (and expensive to compute), how about adding a `z_buffer` parameter to enlarge the vehicle footprint on the z axis ? By increasing this buffer, a user could compensate for inclined roads.

By the way, did you profile the current version ? I am curious about the performance difference between checking for collision in 2D and 3D.

lchojnack commented:
@maxime-clem Thank you very much for your review. 
 
> I think the main remaining issue is that the polyhedrons are not rotated to match the road inclination. Since this is probably difficult to implement (and expensive to compute), how about adding a `z_buffer` parameter to enlarge the vehicle footprint on the z axis ? By increasing this buffer, a user could compensate for inclined roads.

I will add `z_buffer` parameter. 

> By the way, did you profile the current version ? I am curious about the performance difference between checking for collision in 2D and 3D.

No, I didn't. I will profile the performance in terms of time computation and provide a report.

maxime-clem commented on planning/obstacle_stop_planner/src/planner_utils.cpp:
```diff
@@ -449,6 +449,32 @@ bool withinPolygon(
+  }
+  boost_polygon.outer().push_back(bg::make<Point2d>(cv_polygon.front().x, cv_polygon.front().y));
+
+  for (size_t j = 0; j < candidate_points_ptr->size(); ++j) {
+    Point2d point(candidate_points_ptr->at(j).x, candidate_points_ptr->at(j).y);
+    if (bg::distance(prev_point, point) < radius || bg::distance(next_point, point) < radius) {
+      if (bg::within(point, boost_polygon)) {
+        if (candidate_points_ptr->at(j).z < z_max && candidate_points_ptr->at(j).z > z_min) {
+          within_points_ptr->push_back(candidate_points_ptr->at(j));
```
```suggestion
  for (const auto & p : *candidate_points_ptr) {
    Point2d point(p.x, p.y);
    if (bg::distance(prev_point, point) < radius || bg::distance(next_point, point) < radius) {
      if (bg::within(point, boost_polygon)) {
        if (p.z < z_max && p.z > z_min) {
          within_points_ptr->push_back(p);
```

WJaworskiRobotec replied:
@maxime-clem We are planning to run the test in AWSIM to confirm that the original issue is solved and record some videos. But to do that we need a bit more time to create a new ITRI binary based on humble instead of galactic. I will be working on it next week.

lchojnack replied:
@maxime-clem 
Thank you for your review. I've applied your hints.

maxime-clem reviewed (commented):
Thank you for the update and for providing profiling results.

I tested and it works as expected. The code looks good and I only have some small suggestions.
I think the last thing to do before merging is update the documentation of the `obstacle_stop_planner` for the new feature. Do you know if some tests (either in simulation or on a real vehicle) are planned to confirm that the original issue (https://github.com/autowarefoundation/autoware.universe/issues/2213) is solved by this PR ?

maxime-clem commented on planning/obstacle_stop_planner/src/node.cpp:
```diff
@@ -388,26 +407,50 @@ void ObstacleStopPlannerNode::searchObstacle(
+          one_step_move_vehicle_polygon, stop_param.stop_search_radius, prev_center_point,
+          next_center_point, slow_down_pointcloud_ptr, collision_pointcloud_ptr, z_axis_min,
+          z_axis_max);
+        auto after_size = collision_pointcloud_ptr->size();
+        if ((after_size - before_size) > 0) {
+          auto obstacle_ros_pointcloud_debug_ptr = std::make_shared<PointCloud2>();
+          pcl::toROSMsg(*collision_pointcloud_ptr, *obstacle_ros_pointcloud_debug_ptr);
+          obstacle_ros_pointcloud_debug_ptr->header.frame_id = trajectory_header.frame_id;
+          pub_point_cloud_->publish(*obstacle_ros_pointcloud_debug_ptr);
```
Should this code be removed ?

lchojnack replied:
@maxime-clem Thank you for your review. I leave this debug code in order to visualize the obstacle points as a point cloud. What do you think about adding a parameter enabling debugging visualization of an obstacle instead of removing it?

maxime-clem replied:
The main issue currently is that the code does not build (`pub_point_cloud_` is not declared).

Otherwise it is fine to leave this code but if it is heavy to run, rather than adding parameter I recommend having a condition on the number of subscribers to the debug topic.
```C++
if(debug_pub_->get_subscription_count() + debug_pub_->get_intra_process_subscription_count() > 0) {
  /* calculate and publish debug pointclould */
}
```

lchojnack replied:
> The main issue currently is that the code does not build (pub_point_cloud_ is not declared).

I'm really sorry that the code does not build. I was checking the code several times, and I did not notice the issue. 

> Otherwise it is fine to leave this code but if it is heavy to run, rather than adding parameter I recommend having a condition on the number of subscribers to the debug topic.

Ok, It is definitely a better idea to publish point cloud depending on registered subscribers to the topic instead add another parameter. I will do it that way.

maxime-clem reviewed (changes requested):
Thanks for the updates. I think there are some debugging code remaining in the code that need to be removed, otherwise everything looks good.
One thing still missing is an update to the `obstacle_stop_planner` documentation but this can be updated in a separate PR.
I am now testing the PR with some scenarios to make sure there are no big issue. Once they finish and once the remaining minor code issues are fixed, I will approve the PR.

maxime-clem reviewed (approved):
The code looks good and the behavior in simulation is as expected.

xmfcx commented:
@lchojnack could you rebase and resolve the conflicts with the `planning/obstacle_stop_planner/src/node.cpp` please? Then we can proceed to merge it.

lchojnack commented:
@xmfcx I've resolved conflicts and I've also tested the implemented functionality. The feature is working as expected. I've tested it using planning simulator. 

I think that would be good to test the feature implemented in PR https://github.com/autowarefoundation/autoware.universe/pull/2647 in order to be sure that during resolving a conflict in `planning/obstacle_stop_planner/src/node.cpp`  the  `chattering prevention mechanism` wasn't broken.

xmfcx commented:
@satoshi-ota could you check if this PR breaks any functionality that was added in your previous PR https://github.com/autowarefoundation/autoware.universe/pull/2647 ?

satoshi-ota commented on planning/obstacle_stop_planner/src/node.cpp:
```diff
@@ -385,16 +405,26 @@ void ObstacleStopPlannerNode::searchObstacle(
+              one_step_move_vehicle_polygon, stop_param.stop_search_radius, prev_center_point,
+              next_center_point, slow_down_pointcloud_ptr, collision_pointcloud_ptr, z_axis_min,
+              z_axis_max)
+          : withinPolygon(
```
@lchojnack @xmfcx  In my understanding, we can choose `withinPolyhedron` or `withinPolygon` by option `enable_z_axis_obstacle_filtering`. So, I think we should add `withinPolyhedron` here.

https://github.com/autowarefoundation/autoware.universe/blob/7ac34852ea37ec3084060004f505bd61698b65f0/planning/obstacle_stop_planner/src/node.cpp#L480-L483

lchojnack replied:
@satoshi-ota Thank you for your review. I've applied the hint and I've added if `enable_z_axis_obstacle_filtering` option to another part of the code.

satoshi-ota replied:
@lchojnack Yes, it is :+1: Thanks :smile:

satoshi-ota commented:
@lchojnack @xmfcx It looks good to me from the point of view of chattering functionality.

xmfcx reviewed (approved):
will resolve the last simple conflicts and merge

lchojnack commented:
@xmfcx Thank you for resolving the last simple conflicts. 

This is just a friendly reminder. This PR should be merged together with parameters PR https://github.com/autowarefoundation/autoware_launch/pull/150

brkay54 commented:
Hello @lchojnack @maxime-clem , sorry to disturb you. I am working on z-axis filtering for predicted objects so I want to ask you that how to arrange the z-axis coordinate of the obstacle in planning simulator? I set the value in dummy objects but it is not working 😞

lchojnack commented:
Hello @brkay54 ,
I've just checked on the latest autoware, and the feature works as expected.

Please make sure:
1. [This](https://github.com/autowarefoundation/autoware_launch/blob/c1da7a1973c31cf8c719eddb9ad847385a04c67a/autoware_launch/config/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/obstacle_stop_planner.param.yaml#L7) parameter is True 
2. There are green box markers on topic 
`/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/debug/marker`
and collision point cloud on topic 
`/planning/scenario_planning/lane_driving/motion_planning/obstacle_stop_planner/debug/collision_pointcloud` as in the picture below
<img width="1985" alt="image" src="https://user-images.githubusercontent.com/8416175/232704911-b195d4cc-28e7-4f60-9d19-713dbfd8073a.png">

I've used the following command to test
`
ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/elevation_test/ vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit
`

brkay54 commented:
Thank you for your answer. If I understood correctly, you used awsim for these tests (setting dummy obstacle). I did not use awsim, just I am using planning simulator. 
![image](https://user-images.githubusercontent.com/45468306/232708371-be7b10a1-f80b-4bf9-aad0-a473b6dc3485.png)

When I set this value to higher, obstacle position on Z axis did not change. 

Also, if you don't mind, can you share your test map with me?

Thank you! @lchojnack

lchojnack commented:
@brkay54 
> If I understood correctly, you used awsim for these tests (setting dummy obstacle).

No, I use planning simulation for testing using the command
`ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/elevation_test/ vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit`

 I just have ready to run 'clean' autoware (without any local changes) on `awsim-stable` branch hash `984e35c`. 
 I use `2D Dummy Bus` for testing, and sometimes I modify its `H vehicle height` as below
 
<img width="579" alt="image" src="https://user-images.githubusercontent.com/8416175/232711034-287f3781-76bc-4577-9201-0dff4d8e7365.png">

> Also, if you don't mind, can you share your test map with me?

Please use the map from [the tutorial](https://autowarefoundation.github.io/autoware-documentation/main/tutorials/ad-hoc-simulation/planning-simulation/). I've just tested and it also works as expected using command
`ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/sample-map-planning vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit`
